lib/spack/spack/package.py

```diff
@@ -719,20 +719,38 @@ def version_urls(self):
         return version_urls
 
     def nearest_url(self, version):
-        """Finds the URL for the next lowest version with a URL.
-           If there is no lower version with a URL, uses the
-           package url property. If that isn't there, uses a
-           *higher* URL, and if that isn't there raises an error.
+        """Returns the nearest URL for a target version.
+
+        The algorithm to compute the URL is the following:
+
+            1. If the target version is associated with a specific URL, that
+            URL will be returned.
+
+            2. Otherwise return the URL associated with the closest higher
+            version.
```

Comment:
I would like to avoid this behavior. It seems like this is the reverse of the current behavior, which is just as non-intuitive.

For a package like:
```python
class UrlOverride(Package):
    homepage = 'http://www.doesnotexist.org'
    url      = 'http://www.doesnotexist.org/url_override-1.0.0.tar.gz'

    version('1.0.0', 'cxyzab')
    version('0.9.0', 'bcxyza', url='http://www.anothersite.org/uo-0.9.0.tgz')
    version('0.8.1', 'cxyzab')
```
I would expect that versions 0.8.1 and 1.0.0 would download from doesnotexist.org, while version 0.9.0 would download from anothersite.org.

If a package does change it's download site in newer releases, this is easy to support with a `url_for_version` function. In my mind, version-specific URL overrides are intended for a single one-off version, and that seems to be the case in the majority of packages that use them in Spack.

Reply:
> If a package does change it's download site in newer releases, this is easy to support with a url_for_version function.

To be fair it's easy to support also a one-off version... Usually I resort to `url_for_version` only when I need more flexibility for weird versioning schemes (e.g. where the version appear formatted in different ways across a single `url`). Anyhow, what you ask for shouldn't be difficult to implement. Waiting for confirmation on how to proceed..

Reply:
I vote for @adamjstewart's suggestion

Reply:
I believe this is also how `git` and `hg` arguments work. If you have a single version that downloads from `git=...`, the versions above and below this version aren't affected.

Reply:
Ready for a second review. Implementing @adamjstewart's suggestion I could get rid completely of the function.